Separate `pattern: $X` in anded patterns to a special field

emjin opened the pull request:
Before getting the matches of a formula, take out any `pattern: $X` or similar that is in a `And` list so that it can be processed separately. When there are other `And` cases, check the metavariable context produced for `$X` instead of getting the matches.

Sorry for the clutter; realized I needed Augmented_ranges to be its own module too late to make a PR for it alone.

Test plan:

`pattern: $X` alone is checked by the python tests. For correctness in the anded case, in `semgrep-core/` run (or `make test`)

```
sc -config tests/OTHER/rules/pattern-x.yaml tests/OTHER/rules/pattern-x.js -lang js
```

The expected results (you can confirm by running semgrep-core) are
```
perf/input/pattern-x.js:17 with rule typescript.react.security.audit.react-css-injection.react-css-injection
             <div style={input}>
perf/input/pattern-x.js:5 with rule typescript.react.security.audit.react-css-injection.react-css-injection
             <div style={input}>
```

To check that this actually worked, you can run

```
sc -config tests/OTHER/rules/pattern-x.js ~/your_root/semgrep/perf/bench/rails/input/rails/actionview/test/ujs/public/vendor/jquery-2.2.0.js -profile -lang js 
```

Before, this will take ~200 seconds, and half of that time will be spent on `uniq_by`. After, this will take ~100 seconds, and only 2 seconds will be spent on `uniq_by`

PR checklist:
- [x] changelog is up to date

emjin committed:
Separate `pattern: $X` in anded patterns to a special field

Before getting the matches of a formula, take out any `pattern: $X` or similar that is in a `And` list so that it can be processed separately. When there are other `And` cases, check the metavariable context produced for `$X` instead of getting the matches.

emjin committed:
Separate `pattern: $X` in anded patterns to a special field

Before getting the matches of a formula, take out any `pattern: $X` or similar that is in a `And` list so that it can be processed separately. When there are other `And` cases, check the metavariable context produced for `$X` instead of getting the matches.

mschwager committed:
Run both normal and manual pre-commit stages in CI (#3419)

* Run both normal and manual pre-commit stages in CI

* Add ruamel 'MappingNode' stub

* Use float, not int

* Fix tests

* Fix 'metric_manager.set_run_timings' target type

emjin committed:
Rename metavariable pattern filter function

emjin committed:
Rename metavariable pattern filter function

emjin committed:
Extract selector patterns

emjin committed:
Extract selector patterns

emjin committed:
Construct result from metavariable binding

emjin committed:
Construct result from metavariable binding

mschwager committed:
Fix some minor test warnings (#3421)

emjin committed:
When there are no other patterns, compute matches

emjin committed:
When there are no other patterns, compute matches

emjin committed:
Add test file

emjin committed:
Add test file

emjin committed:
Add a file for pattern-x

emjin committed:
Add a file for pattern-x

emjin committed:
Move pattern-x to config tests

emjin committed:
Move pattern-x to config tests

mjambon reviewed (approved):
One thing I would do is start each file with a comment explaining what it's for.

IagoAbal commented:
> pattern: $X alone is checked by the python tests.

I am not sure what this means

IagoAbal commented on semgrep-core/src/engine/Augmented_range.mli:
```diff
@@ -0,0 +1,21 @@
+type range_kind = Plain | Inside | Regexp [@@deriving show]
```
It would be good to add more doc to the module interfaces.

IagoAbal commented on semgrep-core/src/engine/Augmented_range.ml:
```diff
@@ -0,0 +1,122 @@
+type range_kind = Plain | Inside | Regexp [@@deriving show]
+
+(* range with metavars *)
+type range_with_mvars = {
```
Once we moved this data type to its own module then I would just call it `t` as it is tradition in OCaml.

aryx replied:
Yep, I agree.
I would call the module maybe Range_with_metavars.ml

IagoAbal commented on semgrep-core/src/engine/Selector.ml:
```diff
@@ -0,0 +1,130 @@
+module R = Rule
```
Should we have an `.mli` for this module?

aryx replied:
Not a super fan of the name, Selector.ml.
Also, at some point we want to generalize this to handle different special cases, not just
pattern:$PROP but also pattern: include("foo") ..., and other things we think we can
specialize.

IagoAbal commented on semgrep-core/src/engine/Selector.ml:
```diff
@@ -0,0 +1,130 @@
+  lazy_matches : RP.times RP.match_result Lazy.t;
+}
+
+type sformula =
```
I would name the module and the type a bit different, something derived from "optimized formula". Looking into the future, I imagine that we will analyze the formula and produce a new "optimized" one. So this would not be specific to the `pattern: $X` case.

aryx replied:
yep, I agree.

aryx replied:
Actually the best would be to move this Optimize_formula.ml or Specialize_formula.ml in src/optimizing/
to keep the engine/ clean.

IagoAbal commented on semgrep-core/src/engine/Selector.ml:
```diff
@@ -0,0 +1,130 @@
+let match_selector (sel_opt : selector option) : AR.ranges =
+  let get_match (x : RP.times RP.match_result) = x.matches in
+  match sel_opt with
+  | None -> failwith "empty And; no positive terms in And"
```
This error message has been taken out of context, so it doesn't make a lot of sense here. Why should we fail with `"empty And; ..."` if `sel_opt` is `None` ? This function could instead return `AR.ranges option` and we could raise the error in `evaluate_formula`.

IagoAbal commented on semgrep-core/src/engine/Selector.ml:
```diff
@@ -0,0 +1,130 @@
+
+let formula_to_sformula match_func formula =
+  let rec formula_to_sformula formula =
+    let selector_from_formula f =
```
This function could be pull out of the `let rec formula_to_sformula`, would make it clear that `selector_from_formula` can be understood separately and does not depend on any of the arguments of `let rec formula_to_sformula`.

aryx replied:
Yep, @emjin it's usually better to avoid nested functions. In the end it makes the code a bit more complicated.

IagoAbal commented on semgrep-core/src/engine/Selector.ml:
```diff
@@ -0,0 +1,130 @@
+            | None, Some s -> (Some s, acc)
+            | Some s1, Some s2 ->
+                if selector_equal s1 s2 then (Some s1, acc) else (None, [])
+            (* This will never be valid *)
```
What do you mean by valid here? Do you mean that the pattern-and is not valid?

Maybe I'm missing something important here, but it seems to me that:
```
patterns:
  - pattern: $X
  - pattern: $Y
```
which should match something will become `(None, [])` which matches nothing?

aryx commented on CHANGELOG.md:
```diff
@@ -22,6 +22,7 @@ This project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html
   under a different language
 
 ### Changed
+- When anded with other patterns, `pattern: $X` will not be evaluated on its own, but will look at the context and find `$X` within the metavariables bound
```
you should add "which should be significantly faster"

aryx reviewed (changes requested):
A few comments to address and we can then merge this.

IagoAbal committed:
cli: Allow users to disable certain matching features in a rule (#3407)

Follows: eb54e024f0c ("Allow users to disable certain matching features in a rule (#3338)")

emjin committed:
Memoize range_of_any for statements (#3422)

* Memoize range_of_any for statements

Save the range of statements so that range_of_any is much faster.

Test plan:

make test

* CHANGELOG

jtimmons committed:
Fixed issue where metavariable comparison was failing in mixed int/float comparisons (#3424)

emjin committed:
Iago's comments

emjin committed:
Merge branch 'EJ-selector-pattern' of github.com:returntocorp/semgrep into EJ-selector-pattern

IagoAbal reviewed (approved):
Great!

emjin commented:
Messed up this branch, see https://github.com/returntocorp/semgrep/pull/3435
